execution.py

```diff
@@ -4,43 +4,115 @@
+            if hasattr(class_def, "IS_CHANGED"):
+                if "is_changed" in node:
+                    self.is_changed[node_id] = node["is_changed"]
+                else:
+                    input_data_all = get_input_data(node["inputs"], class_def, node_id, self.outputs_cache)
+                    try:
+                        is_changed = map_node_over_list(class_def, input_data_all, "IS_CHANGED")
+                        node["is_changed"] = [None if isinstance(x, ExecutionBlocker) else x for x in is_changed]
+                        self.is_changed[node_id] = node["is_changed"]
+                    except:
+                        node["is_changed"] = float("NaN")
+                        self.is_changed[node_id] = node["is_changed"]
+            else:
+                self.is_changed[node_id] = False
+        return self.is_changed[node_id]
+
+class CacheSet:
+    def __init__(self, lru_size=None):
+        if lru_size is None or lru_size == 0:
+            self.init_classic_cache() 
+        else:
+            self.init_lru_cache(lru_size)
+        self.all = [self.outputs, self.ui, self.objects]
+
+    # Useful for those with ample RAM/VRAM -- allows experimenting without
+    # blowing away the cache every time
+    def init_lru_cache(self, cache_size):
+        self.outputs = LRUCache(CacheKeySetInputSignature, max_size=cache_size)
+        self.ui = LRUCache(CacheKeySetInputSignatureWithID, max_size=cache_size)
+        self.objects = HierarchicalCache(CacheKeySetID)
+
+    # Performs like the old cache -- dump data ASAP
+    def init_classic_cache(self):
+        self.outputs = HierarchicalCache(CacheKeySetInputSignature)
+        self.ui = HierarchicalCache(CacheKeySetInputSignatureWithID)
+        self.objects = HierarchicalCache(CacheKeySetID)
+
+    def recursive_debug_dump(self):
+        result = {
+            "outputs": self.outputs.recursive_debug_dump(),
+            "ui": self.ui.recursive_debug_dump(),
+        }
+        return result
 
-def get_input_data(inputs, class_def, unique_id, outputs={}, prompt={}, extra_data={}):
+def get_input_data(inputs, class_def, unique_id, outputs=None, prompt={}, dynprompt=None, extra_data={}):
     valid_inputs = class_def.INPUT_TYPES()
     input_data_all = {}
     for x in inputs:
         input_data = inputs[x]
-        if isinstance(input_data, list):
+        input_type, input_category, input_info = get_input_info(class_def, x)
+        if is_link(input_data) and (not input_info or not input_info.get("rawLink", False)):
             input_unique_id = input_data[0]
             output_index = input_data[1]
-            if input_unique_id not in outputs:
-                input_data_all[x] = (None,)
+            if outputs is None:
+                continue # This might be a lazily-evaluated input
+            cached_output = outputs.get(input_unique_id)
+            if cached_output is None:
+                continue
```

Comment:
Another good catch.

This was a change in behavior specifically for custom `VALIDATION_INPUTS` functions and custom `IS_CHANGED` functions. We probably didn't see reports of this issue previously just because those two functions aren't frequently used.

Fixing this also changes the behavior of the new `check_lazy_status` function and unevaluated lazy arguments. (Previously, those had to be declared as optional function arguments with `= None`.)